File: typescript/example/env.ts
import { existsSync, readFileSync } from 'node:fs';
import { dirname, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';

export function loadExampleEnv(moduleUrl: string): void {
  for (const path of [
    resolve(process.cwd(), '.env'),
    resolve(dirname(fileURLToPath(moduleUrl)), '.env'),
  ]) {
    loadDotenvFile(path);
  }
}

function loadDotenvFile(path: string): void {
  if (!existsSync(path)) return;

  for (const rawLine of readFileSync(path, 'utf8').split(/\r?\n/)) {
    let line = rawLine.trim();
    if (!line || line.startsWith('#')) continue;
    if (line.startsWith('export ')) line = line.slice('export '.length);
    const equals = line.indexOf('=');
    if (equals < 0) continue;

    const key = line.slice(0, equals).trim();
    if (!key || process.env[key]) continue;
    process.env[key] = parseDotenvValue(line.slice(equals + 1));
  }
}

function parseDotenvValue(value: string): string {
  value = value.trim();
  if (value.length >= 2) {
    const first = value[0];
    const last = value[value.length - 1];
    if ((first === '"' && last === '"') || (first === "'" && last === "'")) {
      return value.slice(1, -1);
    }
  }
  return value;
}

File: python/example/env.py
from __future__ import annotations

import os
from pathlib import Path


def load_example_env() -> None:
    for path in (Path.cwd() / ".env", Path(__file__).with_name(".env")):
        load_dotenv_file(path)


def load_dotenv_file(path: Path) -> None:
    try:
        lines = path.read_text().splitlines()
    except OSError:
        return

    for raw_line in lines:
        line = raw_line.strip()
        if not line or line.startswith("#"):
            continue
        if line.startswith("export "):
            line = line[len("export ") :]
        key, sep, value = line.partition("=")
        if not sep:
            continue
        key = key.strip()
        if not key or os.environ.get(key):
            continue
        os.environ[key] = parse_dotenv_value(value)


def parse_dotenv_value(value: str) -> str:
    value = value.strip()
    if len(value) >= 2 and value[0] == value[-1] and value[0] in {"'", '"'}:
        return value[1:-1]
    return value
